sw/device/tests/entropy_src_fw_override_test.c

```diff
@@ -0,0 +1,407 @@
+  // This value is arbitrary, it could be 1 but since there is some
+  // overhead in dif_entropy_src_observe_fifo_nonblocking_read, it's better
+  // to read several words every time to drain the FIFO quickly.
+  const size_t kDrainCount = 32;
+  size_t len;
+  // Read from the FIFO until we get a short read which means that the FIFO was
+  // emptied.
+  LOG_INFO("drain observe FIFO overflow...");
+  do {
+    len = kDrainCount;
+    CHECK_DIF_OK(dif_entropy_src_observe_fifo_nonblocking_read(&entropy_src,
+                                                               NULL, &len));
+  } while (len == kDrainCount);
+}
+
+void ottf_external_isr(uint32_t *exc_info) {
+  dif_rv_plic_irq_id_t plic_irq_id;
+  CHECK_DIF_OK(dif_rv_plic_irq_claim(&rv_plic, kTopEarlgreyPlicTargetIbex0,
+                                     &plic_irq_id));
+
+  top_earlgrey_plic_peripheral_t peripheral =
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralEntropySrc);
+
+  // If the observe buffer overflows while we're still collecting samples
+  // then we need to drain it and start again or our samples won't be
+  // contiguous.
+  if (words_to_input > 0 && entropy_src_fifo_has_overflowed()) {
+    drain_observe_fifo();
+    // Reset the input buffer.
+    input_buf -= (kFifoBufferSizeWords - words_to_input);
+    words_to_input = kFifoBufferSizeWords;
+  }
+
+  if (words_to_input > 0) {
+    // Move at most 32 samples into the input buffer.
+    size_t len = words_to_input;
+    if (len > 32) {
+      len = 32;
+    }
+
+    CHECK_DIF_OK(dif_entropy_src_observe_fifo_nonblocking_read(
+        &entropy_src, input_buf, &len));
+
+    input_buf += len;
+    words_to_input -= len;
+  }
+
+  // If we've filled the input buffer and drained the output buffer, swap
+  // them.
+  if (words_to_input == 0 && words_to_output == 0) {
+    // Swap the buffers.
+    uint32_t *output_tmp = output_buf;
+    output_buf = input_buf - kFifoBufferSizeWords;
+    input_buf = output_tmp - kFifoBufferSizeWords;
+
+    words_to_input = kFifoBufferSizeWords;
+    words_to_output = kFifoBufferSizeWords;
+  }
+
+  if (words_to_output > 0) {
+    // Move out at most 16 samples into the override FIFO.
+    // This number has to be small otherwise the conditioner runs for too long
+    // and we spend too much time in this ISR.
+    size_t len = words_to_output;
+    if (len > 16) {
+      len = 16;
+    }
+
+    // Start the SHA3 block ready for us to start inserting into the FIFO.
+    CHECK_DIF_OK(dif_entropy_src_fw_override_sha3_start_insert(
+        &entropy_src, kDifToggleEnabled));
+
+    size_t words_written = 0;
+    CHECK_DIF_OK(dif_entropy_src_fw_ov_data_write(&entropy_src, output_buf, len,
+                                                  &words_written));
+
+    output_buf += words_written;
+    words_to_output -= words_written;
+
+    // Disable the SHA3 block, allowing it to finish processing and pass the
```

Comment:
nit: I don't think this technically disables the SHA3 block, rather as you wrote it triggers the SHA3 conditioner (and disable the insertion mecanism).